bin/util/transifex.js

```diff
@@ -0,0 +1,574 @@
+          logThenThrow(value, 'invalid full property');
+        return pluralForms[0];
+      }));
+    }
+    return value;
+  };
+  const messages = parse(
+    fs.readFileSync(`${localesDir}/${sourceLocale}.json`).toString(),
+    reviver
+  );
+
+  // Read the component messages.
+  messages.component = {};
+  for (const [componentName, filename] of filenamesByComponent) {
+    const content = fs.readFileSync(filename).toString();
+    const match = content.match(/<i18n( +lang="json5")? *>/);
+    if (match != null) {
+      const begin = match.index + match[0].length;
+      const end = content.indexOf('</i18n>', begin);
+      if (end === -1) logThenThrow(filename, 'invalid single file component');
+      // Trimming so that if there is an error, the line number is clear.
+      const json = content.slice(begin, end).trim();
+      try {
+        messages.component[componentName] = parse(json, reviver)[sourceLocale];
+      } catch (e) {
+        // eslint-disable-next-line no-console
+        console.error(`could not parse the Vue I18n JSON of ${componentName}`);
+        throw e;
+      }
+    }
+  }
+
+  return messages;
+};
+
+
+
+////////////////////////////////////////////////////////////////////////////////
+// WRITE TRANSLATIONS
+
+// Stores a source message with the corresponding translation.
+class Translation {
+  constructor(parent, key) {
+    this.parent = parent;
+    this._key = key;
+    this.source = parent._source[key];
+  }
+
+  get root() { return this.parent.root; }
+
+  // The translation is "live": if it changes in the parent, it will change here
+  // as well.
+  get translated() { return this.parent._translated[this._key]; }
+
+  toJSON(key) {
+    if (this.translated.isEmpty()) return undefined;
+    return key === 'full' && this.translated.length !== 1
+      ? Array.from(this.translated)
+      : this.translated.toVueI18n();
+  }
+}
+
+// `Translations` stores an object of source messages with the corresponding
+// translations. It provides methods to modify the translations.
+class Translations {
+  constructor(parent, key, source, translated) {
+    this.parent = parent;
+    this._key = key;
+    if (source == null || typeof source !== 'object')
+      logThenThrow(source, 'invalid source');
+    this._source = source;
+    if (translated == null || typeof translated !== 'object')
+      logThenThrow(translated, 'invalid translated');
+    this._translated = translated;
+    this.size = Object.keys(source).length;
+  }
+
+  get root() { return this.parent == null ? this : this.parent.root; }
+
+  // Returns a translation or another set of translations. In either case, the
+  // result is "live": a change to this node will be reflected in its children.
+  get(key) {
+    const sourceValue = this._source[key];
+    if (sourceValue == null) return undefined;
+    if (sourceValue instanceof PluralForms) return new Translation(this, key);
+    if (this._translated[key] == null) this._translated[key] = {};
```

Comment:
I am terrified by a getter that potentially mutates!

Reply:
Ah interesting, I hadn't thought of that. I'm going to throw out a few different thoughts — let me know what you think.

- Here we're mutating internal data within the `Translations` object, but we're not mutating anything externally facing: any descendant `Translation` object will have a `translated` property that is `null`. `this._translated` was passed to the constructor, but `Translations` assumes control of `source` and `translated` once it's passed them.
- Setting `this._translated` here means that `Translations` can always assume that `this._translated` is an object, avoiding a `null` check in a few places.
- Setting `this._translated` is one way to ensure that parent and child `Translations` stay in sync. If `this._source` is an object, and `this._translated` is `null`, then that means that `this._source` is an object that only contains linked locale messages (or other objects that only contain linked locale messages). Eventually we add those linked locale messages to `this._translated` using `set()`, but when doing so, we want to ensure that the parent `Translations` is aware of each addition. We could do so in `set()` instead, but that would require mutating `this.parent._translations`:

```diff
 get(key) {
   const sourceValue = this._source[key];
   if (sourceValue == null) return undefined;
   if (sourceValue instanceof PluralForms) return new Translation(this, key);
-  if (this._translated[key] == null) this._translated[key] = {};
-  return new Translations(this, key, sourceValue, this._translated[key]);
+  return new Translations(
+    this,
+    key,
+    sourceValue,
+    this._translated != null ? this._translated[key] : null
+  );
 }
 
 set(key, translated) {
   if (!(this._source[key] instanceof PluralForms))
     logThenThrow(key, 'invalid key');
   if (!(translated instanceof PluralForms))
     logThenThrow(translated, 'invalid translated');
+
+  const nodes = [];
+  for (let node = this; node != null && node._translated == null; node = node.parent) {
+    nodes.push(node);
+  while (nodes.length !== 0) {
+    const node = nodes.pop();
+    node._translated = {};
+    if (node.parent != null)
+      node.parent._translated[node._key] = node._translated;
+  }
+
   this._translated[key] = translated;
   return this;
 }
 ```

Happy to add a comment about all this or to update `set()`. Or let me know if there's a better way to do it!

Reply:
If this were code we were planning to build on, I'd like to spend more time thinking through alternatives. Given that it's probably (hopefully?) not going to need much touching in the future, it's probably ok. All of this work is pretty mutation-heavy.